src/chunkserver/datastore/file_pool.cpp

```diff
@@ -244,24 +251,187 @@ bool FilePool::CheckValid() {
+        LOG(ERROR) << "Open file failed: " << chunkpath;
+        return false;
+    }
+
+    int fd = ret;
+    auto defer = [&](...){ fsptr_->Close(fd); };
+    std::shared_ptr<void> _(nullptr, defer);
+
+    uint64_t chunklen = poolOpt_.fileSize + poolOpt_.metaPageSize;
+    if (onlyMarked) {
+        ret = fsptr_->Fallocate(fd, FALLOC_FL_ZERO_RANGE, 0, chunklen);
+        if (ret < 0) {
+            LOG(ERROR) << "Fallocate file failed: " << chunkpath;
+            return false;
+        }
+    } else {
+        int nbytes;
+        uint64_t nwrite = 0;
+        uint64_t ntotal = chunklen;
+        uint32_t bytesPerWrite = poolOpt_.bytesPerWrite;
+
+        char *buffer = new (std::nothrow) char[bytesPerWrite];
+        if (nullptr == buffer) {
+            LOG(ERROR) << "Allocate buffer failed for cleaning chunk: "
+                       << chunkpath;
+            return false;
+        }
+        memset(buffer, 0, bytesPerWrite);
+
+        while (nwrite < ntotal) {
+            nbytes = fsptr_->Write(fd, buffer, nwrite,
+                std::min(ntotal - nwrite, (uint64_t)bytesPerWrite));
+            if (nbytes < 0) {
+                LOG(ERROR) << "Write file failed: " << chunkpath;
+                return false;
+            } else if (fsptr_->Fsync(fd) < 0) {
+                LOG(ERROR) << "Fsync file failed: " << chunkpath;
+                return false;
+            }
+
+            cleanThrottle_.Add(false, bytesPerWrite);
+            nwrite += nbytes;
+        }
+    }
+
+    std::string targetpath = chunkpath + kCleanChunkSuffix_;
+    ret = fsptr_->Rename(chunkpath, targetpath);
+    if (ret < 0) {
+        LOG(ERROR) << "Rename file failed: " << chunkpath;
+        return false;
+    }
+
+    return true;
+}
+
+bool FilePool::CleaningChunk() {
+    auto popBack = [this](std::vector<uint64_t>* chunks,
+        uint64_t* chunksLeft) -> uint64_t {
+        std::unique_lock<std::mutex> lk(mtx_);
+        if (chunks->empty()) {
+            return 0;
+        }
+
+        uint64_t chunkid = chunks->back();
+        chunks->pop_back();
+        (*chunksLeft)--;
+        currentState_.preallocatedChunksLeft--;
+        return chunkid;
+    };
+
+    auto pushBack = [this](std::vector<uint64_t>* chunks,
+        uint64_t chunkid, uint64_t* chunksLeft) {
+        std::unique_lock<std::mutex> lk(mtx_);
+        chunks->push_back(chunkid);
+        (*chunksLeft)++;
+        currentState_.preallocatedChunksLeft++;
+    };
+
+    uint64_t chunkid = popBack(&dirtyChunks_, &currentState_.dirtyChunksLeft);
+    if (0 == chunkid) {
+        return false;
+    }
+
+    // Fill zero to specify chunk
+    if (!CleanChunk(chunkid, false)) {
+        pushBack(&dirtyChunks_, chunkid, &currentState_.dirtyChunksLeft);
+        return false;
+    }
+
+    LOG(INFO) << "Clean chunk success, chunkid: " << chunkid;
+    pushBack(&cleanChunks_, chunkid, &currentState_.cleanChunksLeft);
+    return true;
+}
+
+void FilePool::CleanWorker() {
```

Comment:
how to clean the dirty chunks already in chunkfilepool?

Reply:
The chunkserver will create a thread to clean dirty chunk when it start, [see code](https://github.com/opencurve/curve/pull/322/files#diff-0802b07f0949ab1a18239a3e3a4e19c2d67f243e31e6465223955a582b1f74b3R393)